docs/architecture/adr-048-consensus-fees.md

```diff
@@ -0,0 +1,169 @@
+
+## Context
+
+Currently, each validator configures it's own `minimal-gas-prices` in `app.yaml`. But setting a proper minimal gas price is critical to protect network from dos attack, and it's hard for all the validators to pick a sensible value, so we proposal to maintain the gas price in consensus level.
+
+Since tendermint 0.35 has supported mempool prioritization, we can take advantage of that to implement more sophisticated gas fee system.
+
+## Multi-Tier Price System
+
+We propose a multi-tier price system on consensus to provide maximum flexibility:
+
+- Tier 1: a constant gas price, which could only be modified occasionally through governance proposal.
+- Tier 2: a dynamic gas price which is adjusted according to previous block load.
+- Tier 3: a dynamic gas price which is adjusted according to previous block load at a higher speed.
+
+The gas price of higher tier should bigger than the lower tier.
+
+The transaction fees are charged with the exact gas price calculated on consensus.
+
+### Transaction Fields
+
+Add a new `fee_tier` field in `Tx` to specify the preferred tier of service.
+
+Change the semantic of `gasPrice` to `gasPriceCap`, which is the maximum gas price can be charged, if the consensus gas price is larger than `gasPriceCap`, the transaction won't be included and will stay in the mempool until the gas price will go to the `gasPriceCap` level. The mempool can eventually prune old transactions.
+
+### Tx Prioritization
+
+Transactions are prioritized based on the tier, the higher the tier, the higher the priority.
+
+Within the same tier/priority, the transactions are processed in FIFO manner.
+
+### `min-gas-prices`
+
+Deprecate the current per-validator `min-gas-prices` configuration, since it would confusing for it to work together with the consensus gas price.
+
+### Adjust For Block Load
+
+For tier 2 and tier 3 transactions, the gas price is adjusted according to previous block load, the logic could be similar to EIP-1559:
+
+```python
+def adjust_gas_price(gas_price, parent_gas_used, tier):
+  if parent_gas_used == tier.parent_gas_target:
+    return gas_price
+  elif parent_gas_used > tier.parent_gas_target:
+    gas_used_delta = parent_gas_used - tier.parent_gas_target
+    gas_price_delta = max(gas_price * gas_used_delta // tier.parent_gas_target // tier.change_speed, 1)
+    return gas_price + gas_price_delta
+  else:
+    gas_used_delta = parent_gas_target - parent_gas_used
+    gas_price_delta = gas_price * gas_used_delta // parent_gas_target // tier.change_speed
+    return gas_price - gas_price_delta
+```
+
+### Block Segment Reservation
+
+Ideally we should reserve block segments for each tier, so the lower tiered transactions won't be completely squeezed out by higher tier transactions, which will force user to use higher tier, and the system degraded to a single tier.
+
+We need help from tendermint to implement that. (TODO Is the current proposed `ABCI++` spec support this usage?)
```

Comment:
The current ABCI++ data type,
https://github.com/tendermint/spec/blob/master/spec/abci%2B%2B/abci%2B%2B_methods_002_draft.md#data-types-introduced-or-modified-in-abci
It needs to check with @sergio-mena

Reply:
up